game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/change/RemoveUnprotectedUnits.java

```diff
@@ -0,0 +1,173 @@
+      }
+    }
+  }
+
+  private boolean attackerHasRetreat(final BattleState.Side side) {
+    return side == BattleState.Side.OFFENSE
+        && (!battleState.getAttackerRetreatTerritories().isEmpty()
+            || battleState.getAttackingUnits().stream().anyMatch(Matches.unitIsAir()));
+  }
+
+  private List<Unit> getAlliedTransports(final GamePlayer player) {
+    final Predicate<Unit> matchAllied =
+        Matches.unitIsTransport()
+            .and(Matches.unitIsNotCombatTransport())
+            .and(Matches.isUnitAllied(player, battleState.getGameData()))
+            .and(Matches.unitIsSea());
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), matchAllied);
+  }
+
+  private Collection<Unit> getAlliedUnits(final GamePlayer player) {
+    final Predicate<Unit> alliedUnitsMatch =
+        Matches.isUnitAllied(player, battleState.getGameData())
+            .and(Matches.unitIsNotLand())
+            .and(Matches.unitIsSubmerged().negate());
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), alliedUnitsMatch);
+  }
+
+  private Collection<Unit> getEnemyUnitsThatCanFire(final GamePlayer player) {
+    final Predicate<Unit> enemyUnitsMatch =
+        Matches.unitIsNotLand()
+            .and(Matches.enemyUnit(player, battleState.getGameData()))
+            .and(Matches.unitIsSubmerged().negate())
+            .and(Matches.unitCanAttack(player));
+    return CollectionUtils.getMatches(battleState.getBattleSite().getUnits(), enemyUnitsMatch);
+  }
+
+  private void checkUnprotectedUnits(final IDelegateBridge bridge, final BattleState.Side side) {
+    // if we are the attacker, we can retreat instead of dying
+    if (attackerHasRetreat(side)) {
+      return;
+    }
+    if (battleState
+        .getUnits(EnumSet.of(BattleState.Side.OFFENSE, BattleState.Side.DEFENSE))
+        .isEmpty()) {
+      return;
+    }
+    final Collection<Unit> unprotectedUnits = getUnprotectedUnits(side);
+    final boolean hasUnitsThatCanRollLeft = areFightingOrSupportingUnitsLeft(side);
+    final boolean enemyHasUnitsThatCanRollLeft =
+        areFightingOrSupportingUnitsLeft(side.getOpposite());
+    if (!hasUnitsThatCanRollLeft && enemyHasUnitsThatCanRollLeft) {
+      battleActions.remove(
+          unprotectedUnits, bridge, battleState.getBattleSite(), side == BattleState.Side.DEFENSE);
+    }
+  }
+
+  private boolean areFightingOrSupportingUnitsLeft(final BattleState.Side side) {
+    final boolean hasUnitsThatCanRollLeft;
+    hasUnitsThatCanRollLeft =
+        battleState.getUnits(EnumSet.of(side)).stream()
+            .anyMatch(
+                getActiveUnits()
+                    .and(
+                        Matches.unitIsSupporterOrHasCombatAbility(
+                            side == BattleState.Side.OFFENSE)));
+    return hasUnitsThatCanRollLeft;
+  }
+
+  private Predicate<Unit> getActiveUnits() {
```

Comment:
This is perhaps too much information for a step class to have, to know exactly when a unit is active. I suspect having `unit` have a method for this would maybe be more appropriate, but failing that, would it make sense to create a `Matches` for this condition?

Reply:
Moving to `Matches` is a good idea but I'm not sure about the name.  The original code called this match `notSubmergedAndType`.  I didn't like that name and thought it looked more like active units.  But if it is going to be in `Matches` it probably needs a better name.

Reply:
I've created a new `Matches.unitIsActiveInTerritory`